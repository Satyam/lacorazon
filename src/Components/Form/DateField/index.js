import React, { useContext, useEffect, useState } from 'react';
import { FormGroup, Label, FormFeedback, FormText, Col } from 'reactstrap';
import { FormikConsumer, ErrorMessage } from 'formik';
import invariant from 'invariant';
import DatePicker from 'react-datepicker';

import 'react-datepicker/dist/react-datepicker.css';
import classNames from 'classnames';

function useFormik() {
  return useContext(FormikConsumer._context);
}

let counter = 0;

export default function DateField({
  name,
  label,
  help,
  className,
  id,
  ...rest
}) {
  invariant(name, 'DateField: name argument is mandatory');

  const {
    values,
    errors,
    touched,
    registerField,
    unregisterField,
    setFieldTouched,
    setFieldValue,
    validationSchema,
  } = useFormik();

  useEffect(() => {
    registerField(name, {
      props: {
        name,
        label,
        help,
        className,
        id,
        ...rest,
      },
    });
    return () =>
      unregisterField(name, {
        props: {
          name,
          label,
          help,
          className,
          id,
          ...rest,
        },
      });
  });

  const [actualId] = useState(id || `F_DF_${counter}`);

  counter = (counter + 1) % Number.MAX_SAFE_INTEGER;

  const invalid = errors[name] && touched[name];

  let actualMin = rest.minDate;
  let actualMax = rest.maxDate;

  if (validationSchema) {
    const tests = validationSchema.fields[name].tests;
    if (!actualMin) {
      const minTest = tests.filter(t => t.TEST_NAME === 'min')[0];
      if (minTest) {
        actualMin = minTest.TEST.params.min;
      }
    }
    if (!actualMax) {
      const maxTest = tests.filter(t => t.TEST_NAME === 'max')[0];
      if (maxTest) {
        actualMax = maxTest.TEST.params.max;
      }
    }
  }
  return (
    <FormGroup row>
      <Label for={actualId} xs={12} lg={2}>
        {label}
      </Label>
      <Col xs={12} lg={8}>
        <DatePicker
          className={classNames('form-control', className, {
            'is-invalid': invalid,
          })}
          name={name}
          id={actualId}
          minDate={actualMin}
          maxDate={actualMax}
          onChange={value => setFieldValue(name, value)}
          onBlur={() => setFieldTouched(name, true)}
          selected={values[name]}
          {...rest}
        />
        {help && <FormText>{help}</FormText>}
        <ErrorMessage
          name={name}
          component={FormFeedback}
          style={{ display: invalid ? 'block' : 'none' }}
        />
      </Col>
    </FormGroup>
  );
}
