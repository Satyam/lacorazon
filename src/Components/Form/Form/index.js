import React from 'react';
import { Form as BSForm, Alert } from 'reactstrap';
import { Formik, Form as KForm } from 'formik';

export default function Form({
  schema,
  values,
  onSubmit,
  children,
  enableReinitialize = true,
  isInitialValid = false,
  onReset,
  validate,
  validateOnBlur = true,
  validateOnChange = true,
  ...rest
}) {
  return (
    <Formik
      validationSchema={schema}
      initialValues={schema ? Object.assign(schema.default(), values) : values}
      onSubmit={(values, formik) => {
        const result = onSubmit(schema ? schema.cast(values) : values, formik);
        if (result && typeof result.then === 'function') {
          result.then(
            () => {
              formik.setSubmitting(false);
            },
            err => {
              formik.setFieldError('*', err);
            }
          );
        }
      }}
      enableReinitialize={enableReinitialize}
      isInitialValid={isInitialValid}
      onReset={onReset}
      validate={validate}
      validateOnBlur={validateOnBlur}
      validateOnChange={validateOnChange}
    >
      {({ errors }) => (
        <BSForm tag={KForm} {...rest}>
          {errors['*'] && <Alert color="danger">{errors['*']}</Alert>}
          {children}
        </BSForm>
      )}
    </Formik>
  );
}
