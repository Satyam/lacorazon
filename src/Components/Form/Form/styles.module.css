.form-border {
  border: #ced4da solid thin;
  padding: 0.5rem;
  margin: 0.5rem;
  border-radius: 0.3rem;
}
