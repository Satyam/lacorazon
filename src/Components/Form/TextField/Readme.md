Shows an `<input>` or `<textarea>` entry box along with the given `label` (if any)
to allow editing of the matching field as provided in the `values` property of
a `<Form>`.

The only required property is `name`

```js
<Form
  values={{
    one: 1
  }}
>
  <TextField name="one" />
</Form>
```

It looks much better with a label.  
In narrower screens, the label will show immediately above the entry box
while in wider screens, it will show to its left.

```js
<Form
  values={{
    one: 1
  }}
>
  <TextField label="Field one" name="one" />
</Form>
```

If `rows` is provided, a `<textarea>` box will be provided

```js
<Form
  values={{
    multi: "I'm a child\nspread over\na few lines"
  }}
>
  <TextField label="Field multi" name="multi" rows={5} />
</Form>
```

Some help text is convenient

```js
<Form
  values={{
    one: 1
  }}
>
  <TextField label="Field one" name="one" help="Just a little help" />
</Form>
```

Some activity

```js
<Form
  values={{
    one: 1
  }}
  onSubmit={values => {
    alert(JSON.stringify(values, null, 2));
  }}
>
  <TextField label="Field one" name="one" />
  <SubmitButton>Accept</SubmitButton>
</Form>
```

Error message (enter some number greater than 5)

```js
<Form
  values={{
    one: 1
  }}
  onSubmit={values => {
    alert(JSON.stringify(values, null, 2));
  }}
>
  <TextField
    label="Field one"
    name="one"
    validate={value => {
      if (Number.isNaN(Number(value))) return 'Just numbers, please';
      if (Number(value) > 5)
        return 'Sorry, not beyond 5. Notice submit button remains disabled';
    }}
  />
  <SubmitButton>Accept</SubmitButton>
</Form>
```
