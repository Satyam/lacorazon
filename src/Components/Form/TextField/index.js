import React, { useContext, useState } from 'react';
import PropTypes from 'prop-types';
import { FormGroup, Label, FormFeedback, FormText, Col } from 'reactstrap';
import { Field as KField, FormikConsumer, ErrorMessage } from 'formik';
import classNames from 'classnames';
import invariant from 'invariant';

function useFormik() {
  return useContext(FormikConsumer._context);
}

let counter = 0;
/**
 * Produces a labeled input box within form
 */
export default function TextField({
  name,
  label,
  id,
  rows,
  help,
  validate,
  ...rest
}) {
  invariant(name, 'TextField: name argument is mandatory');

  const { errors, touched, validationSchema } = useFormik();
  const invalid = errors[name] && touched[name];
  const [actualId] = useState(id || `F_TF_${counter}`);
  counter = (counter + 1) % Number.MAX_SAFE_INTEGER;

  return (
    <FormGroup row>
      <Label for={actualId} xs={12} lg={2}>
        {label}
      </Label>
      <Col xs={12} lg={8}>
        <KField
          component={rows ? 'textarea' : 'input'}
          className={classNames('form-control', {
            'is-invalid': invalid,
          })}
          rows={rows}
          name={name}
          id={actualId}
          validate={
            validate
              ? value =>
                  validate(
                    validationSchema
                      ? validationSchema.fields[name].cast(value)
                      : value
                  )
              : void 0
          }
          {...rest}
        />
        {help && <FormText>{help}</FormText>}
        <ErrorMessage name={name} component={FormFeedback} />
      </Col>
    </FormGroup>
  );
}

TextField.propTypes = {
  /**
   * Name of the field this component is to show or edit.
   * It should match the key of one of the values in the
   * enclosing [`Form`](#form) `values` property.
   */
  name: PropTypes.string.isRequired,
  /**
   * The label to be shown to the user along the input box
   */
  label: PropTypes.string,
  /**
   * An optional `id` attribute.
   * If none is provided, the component will generate a unique one
   * so the `<label for="inputId">` element can match it
   */
  id: PropTypes.string,
  /**
   * If provided, a `<textarea>` will be produced with as many rows
   * as given, instead of a regular `<input>`box
   */
  rows: PropTypes.number,
  /**
   * An optional help text to be shown below the input field
   */
  help: PropTypes.string,
  /**
   * A validation function.
   * It will be called with the value of the field and
   * it should return a text with an error message, if any,
   * or nothing if Ok.
   */
  validate: PropTypes.func,
  /**
   * Any other properties will be passed on to the `<input>` or `<textarea>` elements
   */
  '...rest': PropTypes.object,
};
