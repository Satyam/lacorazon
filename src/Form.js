import React, { useContext } from 'react';
import {
  Form as BSForm,
  FormGroup,
  Label,
  FormFeedback,
  FormText,
  Alert,
  Col,
  Button
} from 'reactstrap';
import { Formik, Form as KForm, Field as KField, FormikConsumer } from 'formik';

function useFormik() {
  return useContext(FormikConsumer._context);
}

export function Form({
  initialValues,
  schema,
  values,
  onSubmit,
  children,
  ...rest
}) {
  return (
    <Formik
      enableReinitialize={true}
      validationSchema={schema}
      initialValues={Object.assign(schema.default(), values)}
      onSubmit={(values, formik) => onSubmit(schema.cast(values), formik)}
      {...rest}
    >
      {({ errors }) => (
        <BSForm tag={KForm}>
          {errors['*'] && <Alert color="danger">{errors['*']}</Alert>}
          {children}
        </BSForm>
      )}
    </Formik>
  );
}

export function TextField({ name, label, rows, help, validate, ...rest }) {
  const { errors, touched, validationSchema } = useFormik();
  const error = errors[name];

  return (
    <FormGroup row>
      <Label for={name} xs={12} lg={2}>
        {label}
      </Label>
      <Col xs={12} lg={8}>
        <KField
          component={rows ? 'textarea' : 'input'}
          className="form-control"
          rows={rows}
          name={name}
          id={name}
          invalid={error && touched[name]}
          validate={
            validate
              ? value => validate(validationSchema.fields[name].cast(value))
              : void 0
          }
          {...rest}
        />
        <FormFeedback>{error}</FormFeedback>
        {help && <FormText>{help}</FormText>}
      </Col>
    </FormGroup>
  );
}

export function SubmitButton({ component: Component = Button, ...rest }) {
  const { isSubmitting, isValid, dirty } = useFormik();
  return (
    <Component
      type="submit"
      disabled={isSubmitting || !isValid || !dirty}
      {...rest}
    />
  );
}
