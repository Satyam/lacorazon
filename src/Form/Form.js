import React from 'react';
import { Form as BSForm, Alert } from 'reactstrap';
import { Formik, Form as KForm } from 'formik';

export default function Form({
  initialValues,
  schema,
  values,
  onSubmit,
  children,
  ...rest
}) {
  return (
    <Formik
      enableReinitialize={true}
      validationSchema={schema}
      initialValues={Object.assign(schema.default(), values)}
      onSubmit={(values, formik) => onSubmit(schema.cast(values), formik)}
      {...rest}
    >
      {({ errors }) => (
        <BSForm tag={KForm}>
          {errors['*'] && <Alert color="danger">{errors['*']}</Alert>}
          {children}
        </BSForm>
      )}
    </Formik>
  );
}
