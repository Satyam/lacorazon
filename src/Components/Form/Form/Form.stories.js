import React, { useContext, useState } from 'react';

import { storiesOf } from '@storybook/react';

import { FormikConsumer } from 'formik';

import Form from '.';
import styles from './styles.module.css';

function useFormik() {
  return useContext(FormikConsumer._context);
}

function ForceValidate() {
  const [done, setDone] = useState(false);
  const { validateForm } = useFormik();
  if (done) return null;
  validateForm();
  setDone(true);
  return null;
}

function ForceSubmit() {
  const [done, setDone] = useState(false);
  const { submitForm } = useFormik();
  if (done) return null;
  submitForm();
  setDone(true);
  return null;
}

storiesOf('Form/Form', module)
  .add('no attributes', () => <Form />)
  .add('with some child', () => <Form>Hello</Form>)
  .add('with some border and child', () => (
    <Form className={styles['form-border']}>Hello</Form>
  ))
  .add('with error', () => (
    <Form
      isInitialValid={false}
      validate={() => ({ '*': 'some made-up error' })}
      values={{}}
    >
      <ForceValidate />
    </Form>
  ))
  .add('with rejected promise validation', () => (
    <Form
      isInitialValid={false}
      onSubmit={() => Promise.reject('some made-up error')}
      values={{}}
    >
      <ForceSubmit />
    </Form>
  ));
