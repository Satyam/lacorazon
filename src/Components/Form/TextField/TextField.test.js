import React from 'react';
import { render, fireEvent, cleanup } from 'react-testing-library';
import * as Yup from 'yup';

import Form from '../Form';
import TextField from '.';

class ErrorBoundary extends React.PureComponent {
  state = { hasError: false };

  static getDerivedStateFromError(err) {
    // Update state so the next render will show the fallback UI.
    return { hasError: err.message };
  }

  // componentDidCatch(err, info) {
  //   this.setState({ hasError: err.message });
  // }

  render() {
    return this.state.hasError || React.Children.only(this.props.children);
  }
}

afterEach(cleanup);

describe('Form/TextField', () => {
  it('should throw with no props as name argument is mandatory', () => {
    const e = console.error;
    console.error = msg => {};
    const { container } = render(
      <ErrorBoundary>
        <Form>
          <TextField />
        </Form>
      </ErrorBoundary>
    );

    expect(container.innerHTML).toMatchSnapshot();
    console.error = e;
  });
  it('should throw with any extra property but name as argument is mandatory', () => {
    const e = console.error;
    console.error = msg => {};
    const { container } = render(
      <ErrorBoundary>
        <Form>
          <TextField label="some label" value="Some value" />
        </Form>
      </ErrorBoundary>
    );
    expect(container.innerHTML).toMatchSnapshot();
    console.error = e;
  });
  it('should validate on field change', () => {
    const validate = jest.fn(() => '');
    const { getByLabelText } = render(
      <Form values={{ one: 1 }}>
        <TextField label="one" name="one" validate={validate} />
      </Form>
    );
    fireEvent.change(getByLabelText('one'), {
      target: { name: 'one', value: '2' },
    });
    expect(validate.mock.calls).toEqual([['2']]);
  });

  it('should generate an id when no id provided', () => {
    const { getByLabelText } = render(
      <Form values={{ one: 1 }}>
        <TextField label="one" name="one" />
      </Form>
    );
    expect(getByLabelText('one').id).toMatch(/^F_TF_\d+$/);
  });

  it('should respect the id provided', () => {
    const { getByLabelText } = render(
      <Form values={{ one: 1 }}>
        <TextField label="one" name="one" id="abcd" />
      </Form>
    );
    expect(getByLabelText('one').id).toBe('abcd');
  });

  it('should cast value to an integer before validate with schema on field change', () => {
    const schema = Yup.object().shape({
      one: Yup.number()
        .integer()
        .truncate()
        .default(0),
    });
    const validate = jest.fn(() => '');
    const { getByLabelText } = render(
      <Form values={{ one: 1 }} schema={schema}>
        <TextField label="one" name="one" validate={validate} />
      </Form>
    );
    fireEvent.change(getByLabelText('one'), {
      target: { name: 'one', value: '2.5' },
    });
    // string '2.5' was transformed into a number 2
    expect(validate.mock.calls).toEqual([[2]]);
  });
});
