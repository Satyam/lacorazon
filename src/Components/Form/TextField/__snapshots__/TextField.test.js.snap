// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form/TextField should throw with any extra property but name as argument is mandatory 1`] = `"TextField: name argument is mandatory"`;

exports[`Form/TextField should throw with no props as name argument is mandatory 1`] = `"TextField: name argument is mandatory"`;
